### DISCUSSION TITLE
Setting metadata for an http URL?

### DISCUSSION THREAD

**Discussion by wez** [svrooij/node-sonos-ts#197] [Q&A] [answered]
Hi!  I'm not actually using this library, I'm doing something with the rust sonor library, but you seem to know things about this stuff, so I'm asking in the hope that you can share some insight.

I'm building a controller daemon that serves up files from my media library via http URLs.
I can queue these up and play them, but the entries later in the queue don't parse out metadata until they are later played by the sonos device.

I was hoping to be able to send in the track metadata at the time they are added to the queue, but when I try that, I get a 500 error with no additional error context. (I'm not sure if the underlying library is hiding that, or whether there is no additional information).

If I ask the sonos device to return the metadata for the current track after assigning one with no metadata, I see something like this:

```xml
<DIDL-Lite xmlns:dc="http://purl.org/dc/elements/1.1/" 
  xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/" 
  xmlns:r="urn:schemas-rinconnetworks-com:metadata-1-0/" 
  xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/">
<item id="-1" parentID="-1" restricted="true">
<res protocolInfo="http-get:*:audio/mpeg:*" duration="0:03:46">http://192.168.1.92:4242/download/library/file/6a3cfe3ac204f15163c5181fdbc24bc958a5f022a3cb9555476441cc21ce2f1a/data.mp3</res>
<r:streamContent></r:streamContent>
<upnp:albumArtURI>/getaa?u=http%3a%2f%2f192.168.1.92%3a4242%2fdownload%2flibrary%2ffile%2f6a3cfe3ac204f15163c5181fdbc24bc958a5f022a3cb9555476441cc21ce2f1a%2fdata.mp3&amp;v=722</upnp:albumArtURI>
<dc:title>Anonymous</dc:title>
<upnp:class>object.item.audioItem.musicTrack</upnp:class>
<dc:creator>Karl Casey</dc:creator>
<upnp:album>Gore</upnp:album>
<upnp:originalTrackNumber>1</upnp:originalTrackNumber>
<r:albumArtist>Karl Casey</r:albumArtist>
</item></DIDL-Lite>
```

If I try and pass exactly that in as metadata, I also get a 500 error.

Any tips of making this sort of thing work? Is this just one of those things where the sonos device won't accept metadata?

**Comment by svrooij** [comment, marked as answer]
I would do:

1. play the file
2. Make it load the metadata
3. Get the metadata from that file
4. Try to set the metadata in the same way

you best guess is to set as minimal as possible, and it’s important that the metadata is encoded inside the xml. So the metadata is an encoded xml.

Here is some information I also got from messing with the speakers. https://sonos.svrooij.io/metadata.html

**Reply by wez** [reply]
> and it’s important that the metadata is encoded inside the xml

Aha! This was it! The library I'm using does no escaping.

Thank you!

**Reply by svrooij** [reply]
And if this really helped you, anything helps: https://github.com/sponsors/svrooij 🤪

You can also use the generator described here to generate all the methods your Sonos speakers have. https://github.com/svrooij/sonos-api-docs/tree/main/generator/sonos-docs
